analyzers/tests/SonarAnalyzer.UnitTest/TestFramework/SymbolicExecution/SymbolicValueAssertions.cs

```diff
@@ -0,0 +1,127 @@
+                .FailWith("Expected {context:SymbolicValue} to have constraint {0}{reason}, but SymbolicValue has no constraints.", expected)
+                .Then
+                .ForCondition(actuals => actuals.Count == 1)
+                .FailWith("Expected {context:SymbolicValue} to have only constraint {0}{reason}, but SymbolicValue has {1} constraints.", _ => expected, x => x.OrderBy(x => x.ToString()))
+                .Then
+                .Given(actuals => actuals[0])
+                .ForCondition(actual => actual == expected)
+                .FailWith("Expected {context:SymbolicValue} to have constraint {0}{reason}, but SymbolicValue has {1} constraint.", _ => expected, actual => actual);
+            return new(this, Subject.AllConstraints.Single());
+        }
+
+        [CustomAssertion]
+        public AndConstraint<SymbolicValueAssertions> HaveNoConstraints(string because = "", params object[] becauseArgs)
+        {
+            Execute.Assertion
+                .BecauseOf(because, becauseArgs)
+                .ForCondition(Subject == null || Subject.AllConstraints.Count() == 0)
+                .FailWith("Expected {context:SymbolicValue} to have no constraints{reason}, but {0} was found.", Subject?.AllConstraints.OrderBy(x => x.ToString()));
+            return new(this);
+        }
+
+        [CustomAssertion]
+        public AndConstraint<SymbolicValueAssertions> HaveOnlyConstraints(params SymbolicConstraint[] expected) =>
+            HaveOnlyConstraints(expected, string.Empty);
+
+        [CustomAssertion]
+        public AndConstraint<SymbolicValueAssertions> HaveOnlyConstraints(IEnumerable<SymbolicConstraint> expected, string because = "", params object[] becauseArgs)
+        {
+            Execute.Assertion
+                .BecauseOf(because, becauseArgs)
+                .ForCondition(expected != null && expected.Any())
+                .FailWith("Expected constraints are empty. Use HaveNoConstraints() instead.")
+                .Then
+                .ForCondition(Subject != null)
+                .FailWith("The {context:SymbolicValue} is null and can not have constraints {0}.", expected)
+                .Then
+                .Given(() => Subject.AllConstraints.ToList())
+                .ForCondition(actual => actual.Count != 0)
+                .FailWith("Expected {context:SymbolicValue} to have constraints {0}{reason}, but SymbolicValue has no constraints.", expected)
+                .Then
+                .Given(actual => new
+                {
+                    missing = expected.Except(actual).OrderBy(x => x.ToString()).ToList(),
+                    additional = actual.Except(expected).OrderBy(x => x.ToString()).ToList()
+                })
+                .ForCondition(given => given is { missing.Count: 0 } or { additional.Count: > 0 })
+                .FailWith("Expected {context:SymbolicValue} to only have constraints {0}{reason}, but constraints {1} are missing. Actual constraints {2}.",
```

Comment:
I find the "only" pointless here, as something is missing. So forbidding the additional doesn't help.
What about removing the 2nd and 3rd "constraints"? This reads just fine:

>Expected sut[symbol] to only have constraints {First, Dummy}, but {Dummy} are missing. Actual are {First}.

```suggestion
                .FailWith("Expected {context:SymbolicValue} to have constraints {0}{reason}, but {1} are missing. Actual are {2}.",
```

Reply:
Also below